mlflow/promptflow/__init__.py

```diff
@@ -0,0 +1,488 @@
+"""
+The ``mlflow.promptflow`` module provides an API for logging and loading Promptflow models.
+This module exports Promptflow models with the following flavors:
+
+Promptflow (native) format
+    This is the main flavor that can be accessed with Promptflow APIs.
+:py:mod:`mlflow.pyfunc`
+    Produced for use by generic pyfunc-based deployment tools and batch inference.
+
+.. _Promptflow:
+    https://microsoft.github.io/promptflow
+"""
+
+import logging
+import os
+import shutil
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+import yaml
+
+import mlflow
+from mlflow import pyfunc
+from mlflow.models import Model, ModelSignature
+from mlflow.models.model import MLMODEL_FILE_NAME
+from mlflow.models.signature import _infer_signature_from_input_example
+from mlflow.models.utils import ModelInputExample, _save_example
+from mlflow.tracking._model_registry import DEFAULT_AWAIT_MAX_SLEEP_SECONDS
+from mlflow.tracking.artifact_utils import _download_artifact_from_uri
+from mlflow.utils.annotations import experimental
+from mlflow.utils.docstring_utils import LOG_MODEL_PARAM_DOCS, format_docstring
+from mlflow.utils.environment import (
+    _CONDA_ENV_FILE_NAME,
+    _CONSTRAINTS_FILE_NAME,
+    _PYTHON_ENV_FILE_NAME,
+    _REQUIREMENTS_FILE_NAME,
+    _mlflow_conda_env,
+    _process_conda_env,
+    _process_pip_requirements,
+    _PythonEnv,
+    _validate_env_arguments,
+)
+from mlflow.utils.file_utils import write_to
+from mlflow.utils.model_utils import (
+    _validate_and_copy_code_paths,
+    _validate_and_prepare_target_save_path,
+)
+from mlflow.utils.requirements_utils import _get_pinned_requirement
+
+_logger = logging.getLogger(__name__)
+
+FLAVOR_NAME = "promptflow"
+
+_MODEL_FLOW_DIRECTORY = "flow"
+_FLOW_ENV_REQUIREMENTS = "python_requirements_txt"
+_UNSUPPORTED_MODEL_ERROR_MESSAGE = (
+    "MLflow promptflow flavor only supports instance defined with 'flow.dag.yaml' file "
+    "and loaded by ~promptflow.load_flow(), found {instance_type}."
+)
+_INVALID_PREDICT_INPUT_ERROR_MESSAGE = (
+    "Input must be a pandas DataFrame with only 1 row "
+    "or a dictionary contains flow inputs key-value pairs."
+)
+_CONNECTION_PROVIDER_CONFIG_KEY = "connection_provider"
+_CONNECTION_OVERRIDES_CONFIG_KEY = "connection_overrides"
+
+
+def get_default_pip_requirements():
+    """
+    Returns:
+        A list of default pip requirements for MLflow Models produced by this flavor.
+        Calls to :func:`save_model()` and :func:`log_model()` produce a pip environment
+        that, at a minimum, contains these requirements.
+    """
+    tools_package = None
+    try:
+        # Note: If user don't use built-in tool in their flow,
+        # then promptflow-tools is not a mandatory dependency.
```

Comment:
Could you double check `promptflow-tools` is really optional? When I tested before reverting, the saving logic depends on the library, while the model itself doesn't depend on it. Please see the description in #10903 for more details.

The model may not depend on it, but the tools is required for logging the model, users environment should always install it hence this check doesn't work.

Reply:
I think if the users' model depend on `promptflow-tools`, `infer_pip_requirements` should capture that (unless any exception happens and fallback to the default list), hence we don't necessarily add it to default list here.

Reply:
Got it, thanks for the patience, checking..

Reply:
@B-Step62 To double confirm, are you testing with the example in examples/promptflow/basic/hello.py? That does require promptflow-tools because it's using a 'prompt' node in the flow which requires promptflow-tools.

Reply:
For the infer_pip_requirments issue, currently promptflow can't detect if user flow depends on promptflow-tools or not, as there are multiple component may depends on promptflow-tools, includes: built-in llm tool, 'prompt' type node, or even custom script with imports from promptflow.tools, we don't have an approach to catch all the requires for now.

Reply:
Example updated: we have remove the promptflow-tools dependency in example.

Reply:
@brynn-code Hmm I tested with the latest example, but it still requires the tool library for loading.

```
PythonLoadError: Failed to load python module from file '/local_disk0/repl_tmp_data/ReplId-48588-ee785-153c7-7/tmp6xepopfq/promptflow_model/flow/hello.py': (ModuleNotFoundError) No module named 'promptflow.tools'
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_core/tool_meta_generator.py:245, in load_python_module_from_file(src_file)
    244 try:
--> 245     spec.loader.exec_module(m)
    246 except Exception as e:
    247     # TODO: add stacktrace to additional info

File <frozen importlib._bootstrap_external>:883, in exec_module(self, module)

File <frozen importlib._bootstrap>:241, in _call_with_frames_removed(f, *args, **kwds)

File /local_disk0/repl_tmp_data/ReplId-48588-ee785-153c7-7/tmp6xepopfq/promptflow_model/flow/hello.py:7
      6 from promptflow import tool
----> 7 from promptflow.tools.common import parse_chat, render_jinja_template
      9 # The inputs section will change based on the arguments of the tool function, after you save the code
     10 # Adding type to arguments and return value will help the system show the types properly
     11 # Please update the function name/signature per need

ModuleNotFoundError: No module named 'promptflow.tools'

The above exception was the direct cause of the following exception:

PythonLoadError                           Traceback (most recent call last)
File <command-3660584937475137>, line 17
     14 with mlflow.start_run():
     15     logged_model = mlflow.promptflow.log_model(flow, artifact_path="promptflow_model")
---> 17 loaded_model = mlflow.pyfunc.load_model(logged_model.model_uri)
     18 print(loaded_model.predict({"text": "Python Hello World!"}))

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/mlflow/pyfunc/__init__.py:681, in load_model(model_uri, suppress_warnings, dst_path, model_config)
    679         model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path, model_config)
    680     else:
--> 681         model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
    682 except ModuleNotFoundError as e:
    683     # This error message is particularly for the case when the error is caused by module
    684     # "databricks.feature_store.mlflow_model". But depending on the environment, the offending
    685     # module might be "databricks", "databricks.feature_store" or full package. So we will
    686     # raise the error with the following note if "databricks" presents in the error. All non-
    687     # databricks moduel errors will just be re-raised.
    688     if conf[MAIN] == _DATABRICKS_FS_LOADER_MODULE and e.name.startswith("databricks"):

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/mlflow/promptflow/__init__.py:454, in _load_pyfunc(path, model_config)
    452 model_flow_path = os.path.join(path, _MODEL_FLOW_DIRECTORY)
    453 model = load_flow(model_flow_path)
--> 454 return _PromptflowModelWrapper(model=model, model_config=model_config)

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/mlflow/promptflow/__init__.py:401, in _PromptflowModelWrapper.__init__(self, model, model_config)
    399 connection_overrides = model_config.get(_CONNECTION_OVERRIDES_CONFIG_KEY, None)
    400 _logger.info("Using connection overrides: %s", connection_overrides)
--> 401 self.model_invoker = FlowInvoker(
    402     self.model,
    403     connection_provider=connection_provider,
    404     connections_name_overrides=connection_overrides,
    405 )

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_sdk/_serving/flow_invoker.py:77, in FlowInvoker.__init__(self, flow, connection_provider, streaming, connections, connections_name_overrides, raise_ex, **kwargs)
     73 # The credential is used as an option to override
     74 # DefaultAzureCredential when using workspace connection provider
     75 self._credential = kwargs.get("credential", None)
---> 77 self._init_connections(connection_provider)
     78 self._init_executor()
     79 self.flow = self.executor._flow

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_sdk/_serving/flow_invoker.py:90, in FlowInvoker._init_connections(self, connection_provider)
     88 connections_to_ignore.extend(self.connections_name_overrides.keys())
     89 # Note: The connection here could be local or workspace, depends on the connection.provider in pf.yaml.
---> 90 connections = get_local_connections_from_executable(
     91     executable=self._executable_flow,
     92     client=PFClient(config={"connection.provider": connection_provider}, credential=self._credential),
     93     connections_to_ignore=connections_to_ignore,
     94     # fetch connections with name override
     95     connections_to_add=list(self.connections_name_overrides.values()),
     96 )
     97 # use original name for connection with name override
     98 override_name_to_original_name_mapping = {v: k for k, v in self.connections_name_overrides.items()}

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_sdk/_utils.py:872, in get_local_connections_from_executable(executable, client, connections_to_ignore, connections_to_add)
    861 def get_local_connections_from_executable(
    862     executable, client, connections_to_ignore: List[str] = None, connections_to_add: List[str] = None
    863 ):
    864     """Get local connections from executable.
    865 
    866     executable: The executable flow object.
   (...)
    869     connections_to_add: The connection names to add when getting connections.
    870     """
--> 872     connection_names = executable.get_connection_names()
    873     if connections_to_add:
    874         connection_names.update(connections_to_add)

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/contracts/flow.py:826, in Flow.get_connection_names(self)
    824     continue
    825 logger.debug(f"Try loading connection names for node {node.name}.")
--> 826 tool = self.get_tool(node.tool) or self._tool_loader.load_tool_for_node(node)
    827 if tool:
    828     node_connection_names = list(self._get_connection_name_from_tool(tool, node).values())

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_core/tools_manager.py:402, in ToolLoader.load_tool_for_node(self, node)
    400     return self.load_tool_for_package_node(node)
    401 elif node.source.type == ToolSourceType.Code:
--> 402     _, tool = self.load_tool_for_script_node(node)
    403     return tool
    404 raise NotImplementedError(f"Tool source type {node.source.type} for python tool is not supported yet.")

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_core/tools_manager.py:445, in ToolLoader.load_tool_for_script_node(self, node)
    438 if not (self._working_dir / path).is_file():
    439     raise InvalidSource(
    440         target=ErrorTarget.EXECUTOR,
    441         message_format="Load tool failed for node '{node_name}'. Tool file '{source_path}' can not be found.",
    442         source_path=path,
    443         node_name=node.name,
    444     )
--> 445 m = load_python_module_from_file(self._working_dir / path)
    446 if m is None:
    447     raise CustomToolSourceLoadError(f"Cannot load module from {path}.")

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-1902b71f-47fc-4ce4-8414-c0ab1acc00bc/lib/python3.10/site-packages/promptflow/_core/tool_meta_generator.py:249, in load_python_module_from_file(src_file)
    246 except Exception as e:
    247     # TODO: add stacktrace to additional info
    248     error_type_and_message = f"({e.__class__.__name__}) {e}"
--> 249     raise PythonLoadError(
    250         message_format="Failed to load python module from file '{src_file}': {error_type_and_message}",
    251         src_file=src_file,
    252         error_type_and_message=error_type_and_message,
    253     ) from e
    254 return m

PythonLoadError: Failed to load python module from file '/local_disk0/repl_tmp_data/ReplId-48588-ee785-153c7-7/tmp6xepopfq/promptflow_model/flow/hello.py': (ModuleNotFoundError) No module named 'promptflow.tools'
```

Reply:
Thanks for your time, really appreciate it, missed two util function import before, I've update and test in a clean environment to ensure it works:
![image](https://github.com/mlflow/mlflow/assets/24237253/cbd80942-dfc5-4805-ac7c-8778a5816f83)